File: abstract_data_type/throttle/c++/driver.cpp
#include<iostream>
#include<cstdlib>

using namespace std;

class throttle {
public:
    void shut_off(){
        position = 0;
    }

    void shift(int amount){
        position += amount;
        if(position < 0)
            position = 0;
        else if(position > 6)
            position = 6;
    }

    double flow() const {
        return (position / 6.0);
    }

    bool is_on() const {
        return (flow() > 0);
    }
private:
    int position;
};

int main(int argc, char* argv[]){
    throttle sample;
    int user_input;
    cout << "I have a throttle with 6 positions." << endl;
    cout << "Where would you like to set the throttle?" << endl;
    cout << "Please type a number from 0 to 6: ";
    cin >> user_input;
    sample.shut_off();
    sample.shift(user_input);
    while(sample.is_on()){
        cout << "The flow is now " << sample.flow() << endl;
        sample.shift(-1);
    }
    cout << "The flow is now off" << endl;
    return EXIT_SUCCESS;
}

File: abstract_data_type/throttle/php/driver.php
<?php

class throttle {
    private int $position;
    public function shut_off(){
        $this->position = 0;
    }

    public function shift(int $amount){
        $this->position += $amount;
        if($this->position < 0)
            $this->position = 0;
        else if($this->position > 6)
            $this->position = 6;
    }

    public function flow(){
        return ($this->position / 6.0);
    }

    public function is_on(){
        return ($this->flow() > 0);
    }
}




// ----------------------------- Script -----------------------------
$sample = new throttle();
$user_input;
echo "I have a throttle with 6 positions.\n";
echo "Where would you like to set the throttle?\n";
echo "Please type a number from 0 to 6:\n";
fscanf(STDIN, "%d\n", $user_input);
$sample->shut_off();
$sample->shift($user_input);
while($sample->is_on()){
    echo "The flow is now ".$sample->flow()."\n";
    $sample->shift(-1);
}
echo "The flow is now off\n";
